docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+Each database table definition results as the creation of an `Entity` by Propel.
+
+**3-tier class hierarchy**: The Propel generated 2-tier `Entity` class hierarchy is injected in the middle with a SCOS module abstract class to enable adding functionality from SCOS module level (see example below).
+
+See [Propel Documentation - Active Record Class](#https://propelorm.org/documentation/reference/active-record.html)
+See more details on domains in [Persistence Schema](#persistence-schema).
+
+**Conventions**
+- `Entity` base classes MUST NOT be defined manually but generated via [Persistence Schema](#persistence-schema). 
+- `Entities` MUST be instantiated and used only from the [Entity Manager](#entity-manager) of the same module.
+- `Entities` MUST NOT leak to any facade's level (as they are heavy, stateful, module specific objects).
+- `Entities` MUST be implemented according to the 3-tier class hierarchy (see in description & example) to support extension from Propel and SCOS.
+- `Entities` MUST use [Query Objects](#query-object) for database operations.
+
+<details><summary markdown='span'>Additional Conventions for Project Development</summary>
+- The 3-tier class hierarchy is NOT applicable for database tables that are independent to SCOS features.
+</details>
+
+**Guidelines**
+- A typical use-case is to define `preSave()` or `postSave()` methods in the `Entity` object.
+- It is recommended to define manager classes instead of overloading the `Entity` with complex or context-specific logic.
+- `Entities` should not leak outside the module's persistence layer.
+
+**Example**
+
+The lowest level class generated by Propel, containing the base Propel functionality. 
+```php
+namespace Orm\Zed\ConfigurableBundle\Persistence\Base;
+
+/**
+ * Base class that represents a row from the 'spy_configurable_bundle_template' table.
+ */
+abstract class SpyConfigurableBundleTemplate implements Propel\Runtime\ActiveRecord\ActiveRecordInterface {...}
+```
+
+The middle level class generated in a Spryker module, containing the SCOS module functionality.
+```php
+namespace Spryker\Zed\ConfigurableBundle\Persistence\Propel;
+
+abstract class AbstractSpyConfigurableBundleTemplate extends Orm\Zed\ConfigurableBundle\Persistence\Base\SpyConfigurableBundleTemplate {...}
+```
+
+The highest level class generated on project, containing the project specific functionality.
+```php
+namespace Orm\Zed\ConfigurableBundle\Persistence;
+
+class SpyConfigurableBundleTemplate extends Spryker\Zed\ConfigurableBundle\Persistence\Propel\AbstractSpyConfigurableBundleTemplate {...}
+```
+
+## Entity Manager
+
+```
+[Organization]
+└── Zed
+    └── [Module]
+        └── Persistence
+            ├── [Module]EntityManagerInterface.php
+            └── [Module]EntityManager.php
+```
+
+**Description**
+
+Persists [Entities](#entity) by using their internal saving mechanism.
```

Comment:
I don't think that we have to use Propel all the time with Entity Managers. Especially where performance is critical and there is a high load of data to be persisted.

```suggestion
Persists [Entities](#entity) either by using their internal saving mechanism, or by using raw SQL statements.
```

Reply:
we should. Propel triggers P&S events and EntityManager is not responsible for it

Reply:
When the data load is high, this pattern does not scale, and cause a lot of problems on production.

We should either suggest a new pattern for high data load, or accept that EntityManager can trigger events when raw SQL is used to persist entities.

Reply:
For core development, we are not ready to consider this freedom. however, i'm going to lift this restriction for project development. i believe it makes 100% sense 👍